# Github repository for
---
# Persistent microbiome members in the common bean rhizosphere: an integrated analysis of space, time, and plant genotype 
---
## By Nejc Stopnisek and Ashley Shade
---
*This work is published*

## To cite this work or code
Stopnisek N and A Shade.  2021.  Persistent microbiome members in the common bean rhizosphere: an integrated analysis of space, time, and plant genotype. The ISME Journal. https://doi.org/10.1038/s41396-021-00955-5  

## Data
The 16S rRNA and ITS amplicon data from the biogeography study (2017) are available on NCBI under the BioProject [PRJNA524532](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA524532). The 16S rRNA amplicon data from the development study (2018) is available on NCBI under BioProject [PRJNA606063](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA606063). The data from Columbian studies for our meta-analysis were obtained using NCBI BioProjects PRJEB26084 and PRJEB19467.

## Abstract
The full potential of managing microbial communities to support plant health is yet-unrealized, in part because it remains difficult to ascertain which members are most important for the plant. However, microbes that consistently associate with a plant species across varied field conditions and over plant development likely engage with the host or host environment. Here, we applied abundance-occupancy concepts from macroecology to quantify the core membership of bacterial/archaeal and fungal communities in the rhizosphere of the common bean (Phaseolus vulgaris). Our study investigated the microbiome membership that persisted over multiple dimensions important for plant agriculture, including major U.S. growing regions (Michigan, Nebraska, Colorado, and Washington), plant development, annual plantings, and divergent genotypes, and also included re-analysis of public data from beans grown in Colombia. We found 48 core bacterial taxa that were consistently detected in all samples, inclusive of all datasets and dimensions. This suggests reliable enrichment of these taxa to the plant environment and time-independence of their association with the plant. More generally, the breadth of ecologically important dimensions included in this work (space, time, host genotype, and management) provides an example of how to systematically identify the most stably-associated microbiome members, and can be applied to other hosts or systems.

## Funding

This work was supported by the [Plant Resilience Institute](https://plantresilience.msu.edu/) at Michigan State University. 

## More Info
[ShadeLab](http://ashley17061.wixsite.com/shadelab/home)
